contracts/interchain-token-service/tests/flow_limit.rs

```diff
@@ -0,0 +1,290 @@
+    approve_gateway_messages(&env, &gateway_client, signers, messages);
+
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+}
+
+#[test]
+fn set_flow_limit_fails_invalid_amount() {
+    let (env, client, _, _, _) = setup_env();
+    let token_id = BytesN::from_array(&env, &[1; 32]);
+
+    let invalid_limit = Some(-1);
+
+    assert_contract_err!(
+        client
+            .mock_all_auths()
+            .try_set_flow_limit(&token_id, &invalid_limit),
+        ContractError::InvalidFlowLimit
+    );
+}
+
+#[test]
+fn flow_limit_resets_after_epoch() {
+    let (env, client, gateway_client, _, signers) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, _) = setup_flow_limit(&env, &client);
+
+    let amount = TEST_FLOW_LIMIT.unwrap();
+
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, amount);
+    approve_gateway_messages(&env, &gateway_client, signers.clone(), messages);
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+    assert_eq!(client.flow_in_amount(&token_id), amount);
+
+    let current_timestamp = env.ledger().timestamp();
+    env.ledger().set_timestamp(current_timestamp + EPOCH_TIME);
+
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, amount);
+    approve_gateway_messages(&env, &gateway_client, signers, messages);
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+    assert_eq!(client.flow_in_amount(&token_id), amount);
+}
+
+#[test]
+fn add_flow_in_succeeds() {
+    let (env, client, gateway_client, _, signers) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, _) = setup_flow_limit(&env, &client);
+
+    let amount = TEST_FLOW_LIMIT.unwrap();
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, amount);
+    approve_gateway_messages(&env, &gateway_client, signers, messages);
+
+    assert_eq!(client.flow_in_amount(&token_id), 0);
+
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+
+    assert_eq!(client.flow_in_amount(&token_id), amount);
+}
+
+#[test]
+#[should_panic(expected = "Error(Contract, #20)")] // ContractError::FlowLimitExceeded
+fn add_flow_in_fails_exceeds_flow_limit() {
+    let (env, client, gateway_client, _, signers) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, _) = setup_flow_limit(&env, &client);
+
+    let amount = TEST_FLOW_LIMIT.unwrap();
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, amount);
+    approve_gateway_messages(&env, &gateway_client, signers.clone(), messages);
+
+    client.execute(&source_chain, &message_id, &source_address, &payload);
```

Comment:
use assert_auth_ok to ensure the test isn't panicking here instead. Otherwise we can't distinguish which execute failed